filestore.go

```diff
@@ -0,0 +1,282 @@
+	if ret != C.TILEDB_OK {
+		return fmt.Errorf("Error exporting file: %s", f.tdbCtx.LastError())
+	}
+
+	return nil
+}
+
+// Import replaces the contents of f with the contents of the named local file.
+// Assumes that the TileDB array exists and is of the appropriate schema.
+func (f *File) Import(name string, mimeType FileStoreMimeType) error {
+	arrayURI := C.CString(f.arrayURI)
+	defer C.free(unsafe.Pointer(arrayURI))
+	fileURI := C.CString(name)
+	defer C.free(unsafe.Pointer(fileURI))
+
+	ret := C.tiledb_filestore_uri_import(f.tdbCtx.tiledbContext, arrayURI, fileURI, C.tiledb_mime_type_t(mimeType))
+	if ret != C.TILEDB_OK {
+		return fmt.Errorf("Error importing file: %s", f.tdbCtx.LastError())
+	}
+
+	return nil
+}
+
+// CreateAndImport creates the array, initializes it with the appropriate schema and imports the file.
+func (f *File) CreateAndImport(name string, mimeType FileStoreMimeType) error {
+	schema, err := CreateFileSchemaFor(f.tdbCtx, name)
+	if err != nil {
+		return err
+	}
+	array, err := NewArray(f.tdbCtx, f.arrayURI)
+	if err != nil {
+		return err
+	}
+	err = array.Create(schema)
+	if err != nil {
+		return err
+	}
+
+	return f.Import(name, mimeType)
+}
+
+// FileReader is an io.Reader for filestore files.
+type FileReader struct {
+	tdbCtx    *Context // the tiledb context for all read operations
+	arrayURI  string   // the uri of the array
+	arraySize int64    // the size of the array as returned by GetSize
+	bytesRead int64    // the total bytes read so far. Used as an offset for read operations
+
+	cArrayURI *C.char // a malloced char* C string for arrayURI. Cached here to avoid malloc on every read.
+}
+
+// NewFileReader returns a FileReader for the array
+// ArrayURI is the uri of an existing array with the filestore schema.
+func NewFileReader(tdbCtx *Context, arrayURI string) (*FileReader, error) {
+	arraySize, err := NewFile(tdbCtx, arrayURI).Size()
+	if err != nil {
+		return nil, err
+	}
+
+	return &FileReader{
+		tdbCtx:    tdbCtx,
+		arrayURI:  arrayURI,
+		arraySize: arraySize,
+		cArrayURI: C.CString(arrayURI),
+	}, nil
+}
+
+// Close releases the resources of reader
+func (r *FileReader) Close() error {
+	C.free(unsafe.Pointer(r.cArrayURI))
+	return nil
+}
+
+// Size returns the uncompressed size of the underlying array.
+func (r *FileReader) Size() int64 {
+	return r.arraySize
+}
+
+// Read satisfies io.Reader.
+func (r *FileReader) Read(p []byte) (n int, err error) {
+	if len(p) == 0 {
+		return 0, nil
+	}
+
+	bytesRemaining := r.arraySize - r.bytesRead
+	if bytesRemaining == 0 {
+		return 0, io.EOF
+	}
+	if int64(len(p)) > bytesRemaining {
+		p = p[0:bytesRemaining]
+	}
+
+	if err := bufferExport(r.tdbCtx, r.cArrayURI, r.bytesRead, p); err != nil {
+		return 0, err
+	}
+	r.bytesRead += int64(len(p))
+
+	return len(p), nil
+}
+
+// FileWriter is an io.Writer for filestore files
+// FileWriter is an inmemory writer. Each Write stores the data in an internal bytes.Buffer.
```

Comment:
Same logic for reader should be extended for the separate writer. This should not need a separate exported type

Reply:
References should also be to io.WriteCloser rather than separate io.Writer and io.Closer